lib/rules/no-mutation-props.js

```diff
@@ -0,0 +1,222 @@
+/**
+ * @fileoverview Prevent mutation of this.props
+ * @author Ian Schmitz, Joey Baker
+ */
+'use strict';
+
+const Components = require('../util/Components');
+const ARRAY_MUTATIONS = ['push', 'pop', 'shift', 'unshift'];
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent mutation of this.props',
+      category: 'Possible Errors',
+      recommended: false
+    }
+  },
+
+  create: Components.detect((context, components, utils) => {
+    /**
+     * Gets a unique ID for a node
+     * @param {Object} node The node to get an ID for
+     */
+    const getId = node => node && node.range.join(':');
```

Comment:
what happens if the range only has one item, and that item is `toString`?

Reply:
I'm not sure I know what you mean? I didn't think it was possible for range to have only one item? This logic is copied from https://github.com/yannickcr/eslint-plugin-react/blob/f1e86b5ba32ed9359eb7823b98f7a1dcc52376e3/lib/util/Components.js#L43-L45

Reply:
Fair, but since it's being used as an object key, without any prefix, there's still the possibility that it could conflict with an Object.prototype key.

Reply:
Sorry, I'm still not following you. Will the AST allow for `range` to be anything other than a tuple of integers?

Reply:
Yes, I'm just not that familiar with what the AST could return. If `range` always contains > 1 item, then this is totally safe when `node` is truthy.

Reply:
Yea, the spec says `range` must be an array with 2 integers: https://eslint.org/docs/developer-guide/working-with-plugins#all-nodes